Child data sets UI

cquammen opened the pull request:
Adds support to the `PipelineModel` for showing child datasets and visualization modules created to visualize them.

It is not yet possible to apply Operators to child datasets, but that's the next item to implement.

cjh1 commented:
@cquammen I ran the Binary Threshold, should I expect to see a child dataset? This is was I see in my UI:

![tomviz](https://cloud.githubusercontent.com/assets/1031053/18168090/fd4ab2e0-7021-11e6-81b1-d80e06abfb45.png)

cquammen commented:
Indeed, you should see the child hanging off the Orthogonal Slice module. It's probably not there because of the JSON file not found issue you are having, which causes the Binary Threshold operator to think it doesn't produce any children.

cquammen commented:
@cjh1 Please try running CMake after updating your local branch. I forgot to add the `BinaryThreshold.json` file to the list of files to copy to the build directory.

cjh1 commented:
@cquammen That resolved the JSON file issue and the child dataset now appears. I have seeing the following signal related message:

```
QObject::connect: invalid null parameter (:0, )

```

I also get a seg fault when I tried to to remove the Binary Threshold, this stack is below:

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff359cbe4 in vtkSMCoreUtilities::GetFileNameProperty (proxy=0x3c47340)
    at /home/cjh/work/source/paraview/ParaViewCore/ServerManager/Core/vtkSMCoreUtilities.cxx:50
50	  if (proxy->GetHints())
(gdb) bt
#0  0x00007ffff359cbe4 in vtkSMCoreUtilities::GetFileNameProperty (proxy=0x3c47340)
    at /home/cjh/work/source/paraview/ParaViewCore/ServerManager/Core/vtkSMCoreUtilities.cxx:50
#1  0x00000000004d9dd1 in tomviz::DataSource::filename (this=0x5a5a020)
    at /home/cjh/work/source/tomviz/tomviz/DataSource.cxx:199
#2  0x000000000051fc02 in tomviz::PipelineModel::data (this=0xd41960, index=..., role=0)
    at /home/cjh/work/source/tomviz/tomviz/PipelineModel.cxx:276
#3  0x00007ffff5abc080 in QStyledItemDelegate::initStyleOption(QStyleOptionViewItem*, QModelIndex const&) const () from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#4  0x00007ffff5abab79 in QStyledItemDelegate::paint(QPainter*, QStyleOptionViewItem const&, QModelIndex const&) const () from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#5  0x00007ffff5a73ea6 in QTreeView::drawRow(QPainter*, QStyleOptionViewItem const&, QModelIndex const&) const () from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#6  0x00007ffff5a7875e in QTreeView::drawTree(QPainter*, QRegion const&) const ()
   from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#7  0x00007ffff5a7db56 in QTreeView::paintEvent(QPaintEvent*) ()
   from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#8  0x00007ffff5827bd8 in QWidget::event(QEvent*) ()
   from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#9  0x00007ffff5924b9e in QFrame::event(QEvent*) ()
   from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#10 0x00007ffff5a4106b in QAbstractItemView::viewportEvent(QEvent*) ()
   from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#11 0x00007ffff5a7ea10 in QTreeView::viewportEvent(QEvent*) ()
   from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Widgets.so.5
#12 0x00007ffff49ef8f3 in QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) ()
   from /home/cjh/Qt5.6.1/5.6/gcc_64/lib/libQt5Core.so.5
#13 0x00007ffff57e4b75 in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()

```

cquammen commented:
Oh, you want to remove the operator ? :smiley: I haven't added proper support for removing operators with children yet, but obviously that's needed. Thanks for the stack trace.
There are additional issues related to removing DataSources reported in #572 that also need to be addressed.

The chances are 100% that you will find issues all over the place. I hate leaving technical debt, but in this case I plan to pay it back quickly. The goal is to get something that Marcus can show next week.

cjh1 commented:
I understand, sorry to trying unsupported operation :smile:

cquammen commented:
I'm glad you did.

So what is supported?

- Loading data
- Adding modules to loaded data sources
- Adding modules to child data sources and changing visualization parameters. Note that the child data source must be selected before adding a visualization module, otherwise it will be added to whichever data source was previously selected.
- Deleting visualization modules from top-level or child data sets

There is also the additional support for returning new data objects from Python scripts via a dictionary, but that is all under the covers.

cryos commented:
I would like to defer supporting operators on child data for now, and get this stuff working/displaying in the UI correctly. Just taking a look at the updated branch, but this is already looking quite large as a pull request so we should get it to the stage where it can be merged before doing much much so that we can see builds on different platforms, etc.

cryos commented:
I am rebuilding ParaView from a clean branch to see if I can get past the freeze I keep seeing locally. The main visual issue I see is that the child data is not a child of the operator, and I don't know how/don't see the vtkTable data showing up in the pipeline at all right now - is that currently manually selecting the TrivialProducer items in the spread sheet view?

Some of these could be carved off for me or Chris to work on, and don't all need to happen in this branch. My clean build of ParaView is not going well, trying a second time as it seems enabling C++11 features results in compile failures right now.

cquammen commented:
> The main visual issue I see is that the child data is not a child of the operator, and I don't know how/don't see the vtkTable data showing up in the pipeline at all right now

Child data as produced by BinaryThreshold should be a child of the last visualization module that is applied to the parent data set.

![image](https://cloud.githubusercontent.com/assets/360056/18174411/7cab2268-703b-11e6-9d30-f5bd8421a84d.png)

Operator results such as the table produced by ConnectedComponents should be a child of the operator.

![image](https://cloud.githubusercontent.com/assets/360056/18174478/b500a160-703b-11e6-88a9-d0ad0ac67b6e.png)

> I don't know how/don't see the vtkTable data showing up in the pipeline at all right now - is that currently manually selecting the TrivialProducer items in the spread sheet view?

Yes, you have to view it manually with this branch. It would be great if you or Chris wanted to implement clicking on the table output in the pipeline viewer and bringing up a spreadsheet view to show it.

cryos commented:
That doesn't look right to me, and I would say that the child data should be a child of the operator too. It then becomes much clearer to me visually that the binary threshold operator produced the data, and logically seems to group the output of that operator such that if I delete the operator I fully expect the outputs of it to go away too.

cryos commented:
What should the connected components statistics end up being called in the spreadsheet view? Not sure if I am missing something, I see an error,

No result named "component_statistics" defined in output dictionary. (:0, )

cquammen commented:
Let's bring in @Hovden who suggested grouping the child data sets after the operators and modules in #392 (look toward the bottom 1/4 of the discussion page).

cryos commented:
Sure, I would offer up as a prime example that we want to work with several operators that have child data, which I see as being vtkTable, vtkImageData, and other derived types. They should all be treated in a fairly common fashion. Having more than one with child vtkImageData results in:

![image](https://cloud.githubusercontent.com/assets/42214/18176714/f0bae23e-7044-11e6-83bd-58653ecd674b.png)

This would be much clearer to me if they were able to have data as a direct child of the operator that produced them. I don't think I see the utility you are aiming for in putting them below the visualization modules.

cquammen commented:
In that use case, I agree with you. However, @Hovden's idea as I understood it was that you could apply only one child data-producing operator, then no more operators could be applied after that. I'm happy to implement whichever way you and Rob decide.

cryos commented:
I think even if we wanted to limit it to that, I would want to pull that "special" final operator down and make the data a child, such that if you were to delete the operator I would fully expect the child data to also disappear, along with all its child modules, etc.

cryos commented:
I am not sure if this branch breaks it, but I am still seeing this error that looks like it might be related to changes made in this branch:

No result named "component_statistics" defined in output dictionary. (:0, )

cquammen commented:
> No result named "component_statistics" defined in output dictionary. (:0, )

Are there other warnings/errors before or after this message? Make sure CMake was run again to copy over the Python files if you haven't.

cryos commented:
CMake was run, but on Unix systems it is symlinking not copying anyway for the very reason that I didn't want to have to deal with ensuring CMake was run each time I changed a Python file. Full output below, but the other things don't look related.

Warning, array does not have Fortran order, making deep copy and fixing...
...done.
No result named "component_statistics" defined in output dictionary. (:0, )
Warning: In /home/marcus/src/ParaView/ParaViewCore/ClientServerCore/Rendering/vtkPVImageSliceMapper.cxx, line 207
vtkPVImageSliceMapper (0x3de6800): Failed to locate selected scalars. Will use image scalars by default.

cquammen commented:
Hmmm, not sure. If you could, can you see what the value of `spreadsheet` is  around line 100 of `ConnectedComponents.py`? It should be a vtkTable object.

cryos commented:
It matches what is in the repo, is there a simple way to check what the keys the C++ side sees are on the OperatorPython code. Right now you just error out, if failing to find the expected key I think it would be useful to dump the keys that are there.

cquammen commented:
Can you print out the value of `spreadsheet` in the Python script when you run it? I'll add some reporting of the keys received to help debug.

cryos commented:
vtkTable (0x8e32550)
  Debug: Off
  Modified Time: 1209709
  Reference Count: 1
  Registered Events: (none)
  Information: 0x4644aa0
  Data Released: False
  Global Release Data: Off
  UpdateTime: 0
  Field Data:
    Debug: Off
    Modified Time: 1209708
    Reference Count: 1
    Registered Events: (none)
    Number Of Arrays: 0
    Number Of Components: 0
    Number Of Tuples: 0
  RowData: 
    Debug: Off
    Modified Time: 1209730
    Reference Count: 1
    Registered Events: (none)
    Number Of Arrays: 3
    Array 0 name = SurfaceArea
    Array 1 name = Volume
    Array 2 name = SurfaceAreaToVolumeRatio
    Number Of Components: 3
    Number Of Tuples: 242
    Copy Tuple Flags: ( 1 1 1 1 1 0 1 1 )
    Interpolate Flags: ( 1 1 1 1 1 0 0 1 )
    Pass Through Flags: ( 1 1 1 1 1 1 1 1 )
    Scalars: (none)
    Vectors: (none)
    Normals: (none)
    TCoords: (none)
    Tensors: (none)
    GlobalIds: (none)
    PedigreeIds: (none)
    EdgeFlag: (none)

cquammen commented:
Thank you. That is what I would expect.

cryos commented:
I think I spotted it - your use of toStdString().c_str() was producing an empty string/NULL here, and so never returned an object. Everything works great when using toLatin1().data() to get a const char* name for the GetItemString call. I can finally view the table too, I usually work with RelWithDebInfo and a very recent GCC, I wonder if either of those was causing the issue. I can make a small pull request with the change.

cryos commented on tomviz/OperatorPython.cxx:
```diff
@@ -227,27 +264,75 @@ bool OperatorPython::applyTransform(vtkDataObject* data)
+
+    // Results (tables, etc.)
+    for (int i = 0; i < m_resultNames.size(); ++i) {
+      const char* name = m_resultNames[i].toStdString().c_str();
```
I think you mean toLatin1 here anyway, and using

  const char* name = m_resultNames[i].toLatin1().data(); 

Gives me the results I would expect locally.

cquammen commented:
Ah, thanks. That would have taken a while for me to find here. I'll make the change as I am making another change in that part of the code to print out the dictionary keys if there is an error.

cryos commented:
Sounds good, looks like this was the root cause of all my issues not being able to see the vtkTable output - that call always gives an empty string/null here.

cquammen commented:
Okay. I'll get out of the habit of calling toStdString() and use toLatin1() instead. Thanks for digging into it with a debugger.

cquammen commented:
I see the errors with the data source in the data after ConnectedComponents is run, know what the problem is, and will address it.

Hovden commented:
I'm available tomorrow if you want to discuss the pipeline over video chat. I think it is a challenging ux problem.

mathturtle commented:
> Okay. I'll get out of the habit of calling toStdString() and use toLatin1() instead.

I'll need to break that habit too.  We may want to look through the existing code for places the wrong one is used.

cryos commented:
I would fix up the clang-format issues, and we can get this merged as-is and see some build results. It is getting close, and we can afford some work in progress in master right now. The more I look at it, and test it out the more convinced I am that the child vtkImageData should be a child of the operation producing it. To have something to demo next week it can appear either way, and I would like to check we are still building binaries correctly once this merges.

cquammen commented:
Implemented, additional commits coming. Here's what it looks like.

![image](https://cloud.githubusercontent.com/assets/360056/18206856/91f05790-70f6-11e6-8a9d-dfbaff99e8f4.png)

I modified ConnectedComponents to produce a child data set for the label map in addition to the table of object properties.

cquammen commented:
Should be set for review and merging

cryos commented:
Looks good to me, merging this is, we can look at other enhancements as we test this out on other platforms. The pipeline can obviously still be changed once merged, but I think this is something that I can talk about next week.

Hovden commented:
OK. I will test things out this evening.
